### PR TITLE
Generic Multipart bodies, upload and download progress

### LINKED ISSUE #193: Upload progress is actually progress of creating a temp file?
In [UploadTaskRequest.kt](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadTaskRequest.kt), the progress callback isn't sending the progress of the file upload, it's incorrectly sending the progress of writing the file to the output stream

Line 42:
```
    //input file data
    if (outputStream != null) {
       FileInputStream(file).use {
           it.copyTo(outputStream, BUFFER_SIZE) { writtenBytes ->
               progressCallback?.invoke(contentLength + writtenBytes, totalLength)
             }
         }
    }
  contentLength += file.length()
  contentLength += write(CRLF)
```

So does that mean when progress is 100%,  that the upload has been submitted to the server, but it doesn't mean that the upload has completed uploading to the remote server?

### LINKED ISSUE #344: How to upload customed Blobs (like DataPart) in multipart way?
I have a Blob list to upload with different names and filenames like

`listOf(blob("a","a1.jpg",size,stream),blob("a","a2.jpg",size,stream),
blob("b","b1.jpg",size,stream),blob("b",b2.jpg",size,stream))`

are there any way to do it with fuel like DataParts?

`DataPart(Blob(),"name", "image/jpeg")`

 or is this right for this situation

```kotlin
fun compareApi(faceImgData: FaceImgData,callback: (Result<CompareCBData, FuelError>)->Unit){
        compare_suffix.httpUpload(Method.POST).blob { request, url ->
            faceImgData.fullImg
        }.name {
            "fullImg"
        }.blobs { request, url ->
            faceImgData.faces
        }.name {
            "faces"
        }.responseObject(CompareCBData.Deserializer()) { request, response, result ->
            callback(result)
        }
    }
```

### LINKED ISSUE #399: multipart/form-data boundary isn't set correctly
```
val result = FuelManager()
  .request(Method.PUT, "/", listOf("lat" to "3"))
  .apply { type = Request.Type.UPLOAD }
  .dataParts { request, url ->
    listOf(DataPart(File("some.jpg")))
  }.awaitStringResult()
```
This made some trouble with ktor server.
```
java.io.IOException: Failed to parse multipart: Content-Type should be multipart/* but it is application/x-www-form-urlencoded
```
[`Request.dataParts`](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Request.kt#L182) doesn't seem set boundary header, and [`UploadTaskRequest.bodyCallBack`](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/core/requests/UploadTaskRequest.kt#L18) uses arbitrary boundary in this situation. It leads to
```
PUT http://localhost/ HTTP/1.1
Accept-Encoding: compress;q=0.5, gzip;q=1.0
Content-Type: application/x-www-form-urlencoded
Cache-Control: no-cache
Pragma: no-cache
User-Agent: Java/1.8.0_172
Host: localhost:8080
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Connection: keep-alive
Content-Length: 209263

--164e202956e
Content-Disposition: form-data; name="lat"
Content-Type: text/plain
```
Content-Type is urlencoded, but body is multipart/form-data, and library user should set boundary manually. Is it intended?

### LINKED ISSUE #416: Fuel.download doesn't handle redirected url correctly
```kotlin
    Fuel.download("https://github.com/kittinunf/Fuel/releases/download/1.12.1/fuel-1.12.1.jar")
            .destination { response, url ->
                File("fuel.jar")
            }.progress { readBytes, totalBytes ->
                val progress = readBytes.toFloat() / totalBytes.toFloat()
                println(progress)
            }.response { req, res, result ->
            }
```

The `redirectResponseInterceptor` added to Fuel.download call a new request `next(request, manager.request(encoding).header(newHeaders).response().second)`. Thus `destination` and `progress` won't work as they are bound to the first request.

### PR DESCRIPTION
**Pull request by SleeplessByte**
> Builds on #485, this time tacking everything `multipart`, meaning `Blob`, `DataPart` and so forth.

> - Extracts request client options to the executor, instead of having them live on request. They only make sense conditionally and are not actually part of the request themselves. We keep growing the list of attributes, and this solves that.
> - Composition for Download and Upload by wrapping request and overriding "this" (by using the "request" property). This allows us to conditionally make functions available and not rely on an enumeration. This also opens up a request that uploads AND downloads.
> - Request cancelability was actually not implemented correctly. Some requests are currently still not cancellable! This change actually only exposes .cancel if the request is.

**EDIT**: Checkboxes indicate which have been extracted out into new PRs.

- [ ] Fixes #399
- [ ] Fixes #344
- [ ] Should allow to implement #463 with an interceptor
- [X] Should allow to implement #464 with an interceptor
- [ ] Adds unknown lengths (blob, datapart)
- [X] Adds generic streams for data parts #479 
- [ ] Add naming blobs via `DataPart.from(blob)` #344 
- [X] Add `request` / `upload` progress for *all* requests #146 
- [X] Fix progress from an original request to a redirected request #416
- [X] Fix progress using temporary byte array #193

### CODE REVIEW

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/RedirectionInterceptor.kt [hunk 1]
```diff
@@ -29,26 +29,29 @@ fun redirectResponseInterceptor(manager: FuelManager) =
...
-                // check whether it is the same host or not
-                if (newUrl.host != request.url.host) {
-                    newHeaders.remove(Headers.AUTHORIZATION)
-                }
+                val encodedRequest = Encoding(
+                    httpMethod = when {
+                        response.statusCode in redirectStatusWithGets -> Method.GET
+                        else -> request.method
+                    },
+                    urlString = when(URI(redirectedUrl).isAbsolute) {
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/RedirectionInterceptor.kt]: **[ktlint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Missing spacing after "when"

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/RedirectionInterceptor.kt [hunk 2]
```diff
@@ -29,26 +29,29 @@ fun redirectResponseInterceptor(manager: FuelManager) =
...
+                        true -> URL(redirectedUrl).toString()
+                        false -> URL(request.url, redirectedUrl).toString()
+                    }
+                ).request
+
+                val newRequest = manager.request(encodedRequest)
+                    .header(
+                        Headers.from(request.headers).let {
+                            when(encodedRequest.url.host != request.url.host) {
```
Comment (kittinunf) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/RedirectionInterceptor.kt]: **[ktlint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
Missing spacing after "when"

File: fuel-coroutines/src/main/kotlin/com/github/kittinunf/fuel/coroutines/Coroutines.kt [hunk 3]
```diff
@@ -118,11 +118,3 @@ suspend fun <U : Any> Request.awaitObjectResult(
     }
     Result.Failure(fuelError)
 }
-
-@Deprecated("please use 'awaitByteArray()'", ReplaceWith("awaitByteArray()", "deserializable"))
-suspend fun Request.awaitResponseResult(): ByteArray = awaitByteArray()
-
-@Deprecated("please use 'awaitObjectResult(deserializable)'", ReplaceWith("awaitObjectResult(deserializable)"))
-suspend fun <U : Any> Request.awaitSafelyObjectResult(
-    deserializable: ResponseDeserializable<U>
```
Comment (markGilchrist) [inline fuel-coroutines/src/main/kotlin/com/github/kittinunf/fuel/coroutines/Coroutines.kt]: do we have a policy on how long we leave deprecated methods in the lib before we remove them?

File: fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt [hunk 4]
```diff
@@ -11,25 +11,25 @@ import com.github.kittinunf.result.Result
 import io.reactivex.Single
 import java.nio.charset.Charset
 
-fun Request.rx_response() = rx_response(ByteArrayDeserializer())
+fun Request.toRxResponse() = toRxResponse(ByteArrayDeserializer())
 
-fun Request.rx_responseString(charset: Charset = Charsets.UTF_8) = rx_response(StringDeserializer(charset))
+fun Request.toRxResponseString(charset: Charset = Charsets.UTF_8) = toRxResponse(StringDeserializer(charset))
```
Comment (markGilchrist) [inline fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt]: great

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/Fuel.kt [hunk 5]
```diff
@@ -30,89 +33,89 @@ class Fuel {
         // get
         @JvmStatic
         @JvmOverloads
-        fun get(path: String, parameters: List<Pair<String, Any?>>? = null): Request =
+        fun get(path: String, parameters: Parameters? = null): Request =
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/Fuel.kt]: I think this is a great idea

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Blob.kt [hunk 6]
```diff
@@ -3,7 +3,7 @@ package com.github.kittinunf.fuel.core
 import java.io.InputStream
 
 data class Blob(
-    val name: String = "",
+    val name: String,
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Blob.kt]: out of interest why did you remove this default arg?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Body.kt [hunk 7]
```diff
@@ -38,28 +37,30 @@ data class DefaultBody(
...
-            openReader = CONSUMED_READER
+            openStream = CONSUMED_READER
+
+            length
         }
     }
 
-    override fun isEmpty() = openReader == null || (length != null && length == 0)
-    override fun isConsumed() = openReader === CONSUMED_READER
+    override fun isEmpty() = openStream == null || (length != null && length == 0)
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Body.kt]: this could be
`override fun isEmpty() = openStream == null || length == 0`

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt [hunk 8]
```diff
@@ -32,26 +29,22 @@ class Encoding(
...
-        ).apply {
-            if (data != null) body(data)
+        RegularRequest(
+            method = method,
+            path = modifiedPath,
+            url = createUrl(modifiedPath),
+            parameters = parameters ?: emptyList(),
+            headers = headerPairs
+        ).let {
+            when (data) {
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: would an `if` be more readable?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.net.URLConnection
+import java.nio.charset.Charset
+
+data class MultipartBody(private val request: MultipartRequest) : Body {
+    private var inputAvailable: Boolean = true
+
+    override fun isConsumed() = !inputAvailable
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt]: could this be a val with a private set?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/RegularRequest.kt [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+     *  that would lead to closed streams being unable to be read. If the size is known, set it before anything else.
+     *
+     * @param openStream [BodySource] a function that yields a stream
+     * @param calculateLength [Number?] size in +bytes+ if it is known
+     * @param charset [Charset] the charset to write with
+     *
+     * @return [Request] the request
+     */
+    override fun body(openStream: BodySource, calculateLength: BodyLength?, charset: Charset): Request {
+        this.body = DefaultBody.from(openStream = openStream, calculateLength = calculateLength, charset = charset)
```
Comment (markGilchrist) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/core/RegularRequest.kt]: great kDocs!

File: fuel-coroutines/src/main/kotlin/com/github/kittinunf/fuel/coroutines/Coroutines.kt (hunk 3 above)
Comment (SleeplessByte) [reply fuel-coroutines/src/main/kotlin/com/github/kittinunf/fuel/coroutines/Coroutines.kt]: I tend to keep them around at least 2 major versions (so since 1.14), but just take the opportunity to remove them as we don't test them anymore

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt (hunk 9 above)
Comment (SleeplessByte) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt]: `isConsumed` always reflects `inputAvailable`. If it's a val, it will never change, right?

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt (hunk 8 above)
Comment (SleeplessByte) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Encoding.kt]: Personally I like the explicit return. I think this should actually be wrapped in a function `applyBody`, then we can use an if.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Blob.kt (hunk 6 above)
Comment (SleeplessByte) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/Blob.kt]: Haven't figured out completely how I think this API should work, but mostly because `name` doesn't exist on request anymore. Explicit naming works much better. There are already a few tests generating `_0` as name, which does not make any sense, to me.

Removing this default, but allowing some sort of "default options" via the request and injecting it via `blob` would work.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt (hunk 9 above)
Comment (markGilchrist) [reply fuel/src/main/kotlin/com/github/kittinunf/fuel/core/MultipartBody.kt]: sorry I meant 
``` kotlin
var isConsumed :Boolean 
    private set
```

**Comment by SleeplessByte** [comment]
Keeping this open for now for tracking purposes. Please don't ❌ close.

This 🙅 won't be applied as a whole at once, but instead be broken up into multiple PRs 🗡

File: fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt [hunk 11]
```diff
@@ -11,25 +11,25 @@ import com.github.kittinunf.result.Result
...
 
-fun <T : Any> Request.rx_responseObject(deserializable: Deserializable<T>) = rx_response(deserializable)
+fun <T : Any> Request.toRxResponseObject(deserializable: Deserializable<T>) = toRxResponse(deserializable)
```
Comment (kittinunf) [inline fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt]: Hmm I thought we have changed to `rxResponseObject` already in master which is good. Why do we change again here with prefixing it with `to`?

File: fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt [hunk 12]
```diff
@@ -11,25 +11,25 @@ import com.github.kittinunf.result.Result
...
 
-private fun <T : Any> Request.rx_response(deserializable: Deserializable<T>): Single<Pair<Response, Result<T, FuelError>>> =
+private fun <T : Any> Request.toRxResponse(deserializable: Deserializable<T>): Single<Pair<Response, Result<T, FuelError>>> =
         rx {
             val (_, response, result) = response(deserializable)
             response to result
         }
 
-fun Request.rx_bytes() = rx_result(ByteArrayDeserializer())
+fun Request.toRxBytes() = toRxResult(ByteArrayDeserializer())
```
Comment (kittinunf) [inline fuel-rxjava/src/main/kotlin/com/github/kittinunf/fuel/rx/RxFuel.kt]: Here too.

**Comment by kittinunf** [comment]
ah just realize that you need to rebase master first I think. So the changes make sense again?

**Comment by iNoles** [comment]
Wow a lot of Conflicts.
